crates/matrix-sdk/src/client/mod.rs

```diff
@@ -437,48 +453,113 @@ impl Client {
+                    }
+                    Err(e) => {
+                        warn!(
+                            "Failed to deserialize `{}` event, skipping event handler.\n\
+                                Deserialization error: {}",
+                            event_type, e,
+                        );
+                    }
+                }
+            })
+        });
 
-        self
+        let handler_id = self.inner.event_handler_counter.fetch_add(1, SeqCst);
+
+        let handle = EventHandlerHandle { handler_id, ev_id: Ev::ID };
+
+        self.inner
+            .event_handlers
+            .write()
+            .await
+            .entry(H::ID)
+            .or_default()
+            .push(EventHandlerWrapper { handler_fn, handle: handle.clone() });
+
+        handle
     }
 
     pub(crate) async fn event_handlers(&self) -> RwLockReadGuard<'_, EventHandlerMap> {
         self.inner.event_handlers.read().await
     }
 
+    /// Remove the event handler associated with the handle.
+    ///
+    /// Note that handlers that remove themselves will still execute
+    /// with events received in the same sync cycle.
+    ///
+    ///  # Arguments
+    ///
+    /// `handle` - The [`EventHandlerHandle`] that is returned when
+    /// registering the event handler with [`Client::register_event_handler`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use futures::executor::block_on;
+    /// # use url::Url;
+    /// # use tokio::sync::mpsc;
+    /// #
+    /// # let homeserver = Url::parse("http://localhost:8080").unwrap();
+    /// #
+    /// use matrix_sdk::{
+    ///     ruma::events::room::member::SyncRoomMemberEvent,
+    ///     Client, event_handler::EventHandlerHandle
+    /// };
+    /// #
+    /// # block_on(async {
+    /// # let client = matrix_sdk::Client::builder()
+    /// #     .homeserver_url(homeserver)
+    /// #     .server_versions([ruma::api::MatrixVersion::V1_0])
+    /// #     .build()
+    /// #     .await
+    /// #     .unwrap();
+    ///
+    /// client
+    ///     .register_event_handler(
+    ///         |ev: SyncRoomMemberEvent, client: Client, handle: EventHandlerHandle| async move {
+    ///             // Common usage: Check arriving Event is the expected one
+    ///             println!("Expected RoomMemberEvent received!");
+    ///             client.remove_event_handler(handle);
+    ///         },
+    ///     )
+    ///     .await;
+    /// # });
+    /// ```
+    pub async fn remove_event_handler(&self, handle: EventHandlerHandle) {
+        let mut event_handlers = self.inner.event_handlers.write().await;
+
+        if let Some(v) = event_handlers.get_mut(&handle.ev_id) {
+            v.retain(|e| -> bool { e.handle.handler_id.ne(&handle.handler_id) });
+
+            if v.is_empty() {
+                event_handlers.remove(&handle.ev_id);
+            }
+        }
+
+        self.inner.event_handler_counter.fetch_sub(1, SeqCst);
```

Comment:
I guess this is what you're talking about regarding handle uniqueness. I have an easy fix for you 😉 

```suggestion
```

Reply:
alrighty :D

Reply:
this will limit us to (2^64)-1 event handlers